RevGrid: Show popup when a revision is not in the revision grid

gerhardol opened the pull request:
Discussed in relation to #9796 

## Proposed changes

If a objectId is expected to be available in the repo (for instance
retrieved with git-rev-parse), but is not included in the grid,
GE will inform the user with a popup instead of just silently ignore.

This is how GE has behaved for a long time, but there were previously less opportunities to 
enter commit sha that were in the repo but not displayed in the grid. 
There are various filters (including path filter) that hides commits and
links are exposed in the side panel and commit info.

All occurrences of `SetSelectedRevision()` are not getting the popup, it is opt-in
even if most occurrences are included. I have tried to verify that there are no popups in 
"automatic" situations and that the ObjectIds are valid Git sha (normally parsed by git-rev-parse or from git-blame etc).

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

Navigate - go to commit with input that is not commitish.

![image](https://user-images.githubusercontent.com/6248932/150697909-d6a34ef9-b268-477e-bc7a-e86d49fcf23e.png)

### After

![image](https://user-images.githubusercontent.com/6248932/150697919-bb3b7540-0feb-413d-9807-2c1bad431caf.png)

![image](https://user-images.githubusercontent.com/6248932/150697963-8ccea5d1-ad6e-4970-9db4-6ccc85b92f40.png)

## Test methodology <!-- How did you ensure quality? -->

Manual review and check of situations the pops should occur in

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2454,7 +2454,7 @@ private void RevisionInfo_CommandClicked(object sender, ResourceManager.CommandE
                     if (found)
                     {
-                        found = RevisionGrid.SetSelectedRevision(revision);
+                        found = RevisionGrid.SetSelectedRevision(revision, showNoRevisionMsg: !AppSettings.ShowFirstParent);
```
`RevisionInfo_CommandClicked()` is called when links in the Commit tab is clicked.

gerhardol commented on GitUI/CommandsDialogs/FormFileHistory.cs:
```diff
@@ -553,7 +553,7 @@ private void Blame_CommandClick(object sender, ResourceManager.CommandEventArgs
                 if (Module.TryResolvePartialCommitId(e.Data, out var objectId))
                 {
-                    RevisionGrid.SetSelectedRevision(objectId);
+                    RevisionGrid.SetSelectedRevision(objectId, showNoRevisionMsg: true);
```
Similar to FormBrowse

gerhardol commented on GitUI/UserControls/BlameControl.cs:
```diff
@@ -502,7 +502,7 @@ private void ActiveTextAreaControlDoubleClick(object sender, EventArgs e)
             {
                 _clickedBlameLine = _lastBlameLine;
-                _revGrid.SetSelectedRevision(_lastBlameLine.Commit.ObjectId);
+                _revGrid.SetSelectedRevision(_lastBlameLine.Commit.ObjectId, showNoRevisionMsg: true);
```
If a blame is not visible in the grid, the "Blame previous revision" and "Blame this revision" are disabled, so the popup should be redundant.

gerhardol commented on GitUI/UserControls/RevisionGrid/RevisionGridMenuCommands.cs:
```diff
@@ -501,11 +500,11 @@ public void GotoCommitExecute()
             if (objectId is not null)
             {
-                _revisionGrid.SetSelectedRevision(objectId);
+                _revisionGrid.SetSelectedRevision(objectId, showNoRevisionMsg: true);
```
`Goto commit` form

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -2454,7 +2454,7 @@ private void RevisionInfo_CommandClicked(object sender, ResourceManager.CommandE
                     if (found)
                     {
-                        found = RevisionGrid.SetSelectedRevision(revision);
+                        found = RevisionGrid.SetSelectedRevision(revision, showNoRevisionMsg: !AppSettings.ShowFirstParent);
```
I think there may be other reasons why a given revision isn't visible, i.e. other filters are applied.

gerhardol replied:
The intention is that the message should be interpreted for any filter not just "advanced filter", branchfilters etc but also show first parents etc.
Clarification?

RussKie replied:
Sorry, the meaning of your message is still escaping me... 

We should avoid using `AppSettings.ShowFirstParent` and other filtering options on their own. All filtering decisions or checks should be done only via `FilterInfo` object (which is essentially an abstraction layer).

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -671,6 +673,11 @@ public bool SetSelectedRevision(ObjectId? objectId, bool toggleSelection = false
                 return true;
             }
 
+            if (showNoRevisionMsg)
```
Why would we not want to show the message?

gerhardol replied:
If the check is done in an "automatic" flow where you do not realize what went wrong.
Also if you for some reason manage to check a complete ObjectId that is not in the repo at all.

Both of these arguments (especially the second one) is possibly not valid scenarios, but I cannot verify all scenarios so this is taking it safe.

gerhardol commented:
Pushed one update. If the revision was not found the selection was still cleared after selecting.
So if the revision grid is filtered and you for instance clicked a branch in the sidepanel that is not visible the selection was just cleared.
Even more confusing when doubleclicking a commit that is not available in Blame (drop down menu is disabled, double click is still possible).

mstv commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -658,8 +658,10 @@ private int FindRevisionIndex(ObjectId? objectId)
+        /// <param name="showNoRevisionMsg">Show popup if revision is not visible in grid. Should only be set if the ObjectId is in the repo.</param>
         /// <returns><c>true</c> if the required revision was found and selected, otherwise <c>false</c>.</returns>
-        public bool SetSelectedRevision(ObjectId? objectId, bool toggleSelection = false)
+        public bool SetSelectedRevision(ObjectId? objectId, bool toggleSelection = false, bool showNoRevisionMsg = false)
```
```suggestion
        public bool SetSelectedRevision(ObjectId? objectId, bool toggleSelection = false, bool showNoRevisionMsg = true)
```
(As you know,) *I* would rather show the message by default and justify the cases where it is suppressed.

gerhardol replied:
I choose the defensive choice: to not change the behavior unless you choose to.
The value is set true almost everywhere.
https://github.com/gitextensions/gitextensions/pull/9833#discussion_r792165847

But we can say that new usage of the code must investigate the behavior and change the defaults.

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -54,6 +57,12 @@ internal MessageBoxes()
         private static MessageBoxes Instance => instance ??= new();
 
+        public static void RevisionFilteredInGrid(IWin32Window? owner, ObjectId objectId)
+            => ShowError(owner, string.Format(Instance._cannotFindRevisionFilter.Text, objectId.ToShortString()), Instance._cannotFindRevisionCaption.Text);
```
In a follow-up PR, we should switch `MessageBoxes` to `TaskDialog` and add an `Action` in order to turn off the filter directly.
Then this `Action` should retry to locate the revision.

gerhardol replied:
I do not think we can disable all possible filters. For instance all commits may not be loaded.
But it could be a suggestion, *Advanced filter* has most common occurrences (and could add more).

RussKie replied:
`FilterInfo` contains all necessary information to know whether filters are applied or not. If there's something missing we need to add those checks there.
If there is no filter applied (as far as `FilterInfo` is concerned) we should show a task dialog with a message "Requested revision cannot be found". Else, if there is a filter applied, we should show a task dialog with an option to reset the filter - if that option is invoked - open the Advanced Filter dialog.

gerhardol replied:
> `FilterInfo` contains all necessary information to know whether filters are applied or not. If there's something missing we need to add those checks there. If there is no filter applied (as far as `FilterInfo` is concerned) we should show a task dialog with a message "Requested revision cannot be found". Else, if there is a filter applied, we should show a task dialog with an option to reset the filter - if that option is invoked - open the Advanced Filter dialog.

 `FilterInfo` is just a subset, there are several others like "--first-parent, --reflog, --simplify-by-decoration" at a quick glance.
Advanced filter will be quite bigger...

Only a few of the uses is from FormBrowse, so the AdvancedFilter is not accissible, an action to reset the filter could be used though.

gerhardol commented:
rebased to resolve merge conflict

mstv reviewed (commented):
+1

gerhardol commented:
> (As you know,) I would rather show the message by default and justify the cases where it is suppressed.

> Why would we not want to show the message?

Changed the defaults to show the popup by defaults.
Also included most of those situations where popups where not set before. I reviewed those again and could not see any action that are automatic. Some of those I did not include earlier are from key board actions and the other is commented as follows:
```
            // Parent-child navigation can expect that SetSelectedRevision is always successful since it always uses first-parents
            _parentChildNavigationHistory = new ParentChildNavigationHistory(objectId => SetSelectedRevision(objectId));
```

` _parentChildNavigationHistory` is invoked at refreshes etc. Maybe this specific occurrence should suppress pops.

A few popup supersessions remains, mostly to avoid double popus.

RussKie commented:
Please correct me if I misunderstand the use case.
I understand it as follows:

A user is searching for a commit by hash.
The hash is not found.
The app shows a warning to the user (e.g., "The requested hash '<hash>'
cannot be found.") with an option to reset current filters, if any applied?
(I.e. FilterInfo)

In which case would we not want to show the warning?

On Thu, 27 Jan 2022, 8:25 am Gerhard Olsson, ***@***.***>
wrote:

> (As you know,) I would rather show the message by default and justify the
> cases where it is suppressed.
>
> Why would we not want to show the message?
>
> Changed the defaults to show the popup by defaults.
> Also included most of those situations where popups where not set before.
> I reviewed those again and could not see any action that are automatic.
> Some of those I did not include earlier are from key board actions and the
> other is commented as follows:
>
>             // Parent-child navigation can expect that SetSelectedRevision is always successful since it always uses first-parents
>             _parentChildNavigationHistory = new ParentChildNavigationHistory(objectId => SetSelectedRevision(objectId));
>
> _parentChildNavigationHistory is invoked at refreshes etc. Maybe this
> specific occurrence should suppress pops.
>
> A few popup supersessions remains, mostly to avoid double popus.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9833#issuecomment-1022617178>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQSX3IQN3MEIHPTMCDUYBRGJANCNFSM5MT2CUXQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
> Please correct me if I misunderstand the use case. I understand it as follows: A user is searching for a commit by hash. The hash is not found. The app shows a warning to the user (e.g., "The requested hash '<hash>' cannot be found.") with an option to reset current filters, if any applied? (I.e. FilterInfo) In which case would we not want to show the warning?

In this scenario, the pop is desired.

But the popup is suppressed for instance in https://github.com/gitextensions/gitextensions/pull/9833/files#diff-613b92e9f0a420f482a14b89d7f9c9d31fb4f0f98e5f343a130136492944f1e1L2457 to avoid two potential popups.

I am not 100% sure about some automated use, see `_parentChildNavigationHistory ` where the command is executed when switching repos. It seems OK.

An example where I do not want a popup (fortunately parent rewriting is used so this is not an issue):

If "goto parent `Ctrl-P`" was implemented with asking Git for the parent for the current selected revision, the sha would be valid but maybe not available in the revision grid. If the implementation first tried the Git direct parent, if that failed asked Git for the "rewritten parent" you would not like to have a popup for the first occurrence.

RussKie commented:
From the layering perspective the RevisionGridControl shouldn't show any
messages, it should only render the supplied content. If there are any
anomalies (e.g., absent hashes) it should throw an exception. Callers must
handle the exception as necessary.

On Fri, 28 Jan 2022, 1:56 am Gerhard Olsson, ***@***.***>
wrote:

> Please correct me if I misunderstand the use case. I understand it as
> follows: A user is searching for a commit by hash. The hash is not found.
> The app shows a warning to the user (e.g., "The requested hash '' cannot be
> found.") with an option to reset current filters, if any applied? (I.e.
> FilterInfo) In which case would we not want to show the warning?
>
> In this scenario, the pop is desired.
>
> But the popup is suppressed for instance in
> https://github.com/gitextensions/gitextensions/pull/9833/files#diff-613b92e9f0a420f482a14b89d7f9c9d31fb4f0f98e5f343a130136492944f1e1L2457
> to avoid two potential popups.
>
> I am not 100% sure about some automated use, see _parentChildNavigationHistory
> where the command is executed when switching repos. It seems OK.
>
> An example where I do not want a popup (fortunately parent rewriting is
> used so this is not an issue):
>
> If "goto parent Ctrl-P" was implemented with asking Git for the parent
> for the current selected revision, the sha would be valid but maybe not
> available in the revision grid. If the implementation first tried the Git
> direct parent, if that failed asked Git for the "rewritten parent" you
> would not like to have a popup for the first occurrence.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9833#issuecomment-1023295212>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRRLGRASKZVJJETKATUYFMJ7ANCNFSM5MT2CUXQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
> From the layering perspective the RevisionGridControl shouldn't show any messages, it should only render the supplied content. If there are any anomalies (e.g., absent hashes) it should throw an exception. Callers must handle the exception as necessary.

It is not expected to be an error if `SetSelectedRevision` fails to find a revision, it returns false. The method is also used in non interactive situations and "two-step" checks. 
Popups cannot be shown in  every situation. 
This PR is not changing the internal handling, just shows the popup in all but a few situations.

RussKie commented:
I'm currently away and don't have access to a computer at the moment. I'd
like to have an opportunity to review this proposal in more details.

On Fri, 28 Jan 2022, 11:11 am Gerhard Olsson, ***@***.***>
wrote:

> From the layering perspective the RevisionGridControl shouldn't show any
> messages, it should only render the supplied content. If there are any
> anomalies (e.g., absent hashes) it should throw an exception. Callers must
> handle the exception as necessary.
>
> It is not expected to be an error if SetSelectedRevision fails to find a
> revision, it returns false. The method is also used in non interactive
> situations and "two-step" checks.
> Popups cannot be shown in every situation.
> This PR is not changing the internal handling, just shows the popup in all
> but a few situations.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9833#issuecomment-1023757102>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVN7V43U4YP3BU4FRTUYHNKJANCNFSM5MT2CUXQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

gerhardol commented:
> I'm currently away and don't have access to a computer at the moment. I'd like to have an opportunity to review this proposal in more details.

@RussKie comments?

RussKie commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:
```diff
@@ -658,22 +658,27 @@ private int FindRevisionIndex(ObjectId? objectId)
-                SetSelectedIndex(index, toggleSelection);
-                _navigationHistory.Push(objectId);
-                return true;
+                if (showNoRevisionMsg)
+                {
+                    MessageBoxes.RevisionFilteredInGrid(this, objectId);
+                }
```
I am very much against having the grid show messages, it doesn't feel appropriate.
This API returns `false` if we couldn't find the requested objectid. The caller (i.e., `FormBrowse`) must then show the error message, if necessary (in the interactive flow).

gerhardol commented:
Squashed, rebased (to resolve merge conflicts) and distributed the popup to all callers.
(Some cleanup too, showing the popup slightly more.)
Note that before #9864 it could be possible to get old commits in the history so popups could occur.

gerhardol commented:
squash merge in one day

mstv reviewed (approved):
Have not run the latest

gerhardol commented:
Minor syntax touchups, removed handling for "gotocommit" that did not work (before this PR, exception occurred), the popup is sufficient.
